### DISCUSSION TITLE
How about a `dirvish-filter` extension?

### DISCUSSION THREAD

**Discussion by alexluigit** [alexluigit/dirvish#58] [Ideas]
> Related to the performance of dired, do you have any ideas that can make dired-filter-groups faster? I like them a lot and would like to have them enabled if I accidentally open a directory with large number of files but the performance is slow. My knowledge of dired internals is non-existent so I don't know if that aspect can be made faster, but if you have any ideas I will like to try in (not so) near future.

Ref: https://github.com/wyuenho/all-the-icons-dired/issues/16

Not that I know of. The reason for that is dired-filter.el heavily relies on dired-revert, which, unfortunately, is very slow especially on large/remote directories. This also puts the filter stack mechanism in a very awkward situation: it's not useful for small directories, yet it's slow on big directories. Besides, this package provides a lot of features that overlaps with dired's inbuilt marking system, of which I'm not a big fan. For the reasons above I've stopped using this package. I'm planning to implement a stripped-down version of this package in the form of a dirvish extension (like dirvish-subtree as to dired-subtree), maybe in the near future.

**Comment by aikrahguzar** [comment]
> Not that I know of. The reason for that is dired-filter.el heavily relies on dired-revert, which, unfortunately, is very slow especially on large/remote directories. 

Why is that? I got lost when I was reading (a while ago) the code in it but the very high level idea seemed to go through the buffer line by line and keep the lines that should be kept which seems like the obvious thing to do. I doesn't seem like that should need dired-revert.

> This also puts the filter stack mechanism in a very awkward situation: it's not useful for small directories, yet it's slow on big directories. Besides, this package provides a lot of features that overlaps with dired's inbuilt marking system, of which I'm not a big fan. For the reasons above I've stopped using this package. I'm planning to implement a stripped-down version of this package in the form of a dirvish extension (like dirvish-subtree as to dired-subtree), maybe in the near future.

This is indeed awkward, but filter stack has been useful to me even in small directories in that I can keep important files at the top. For example I used it to keep tex files at the top and then pdf files and then style files etc in a dired buffer which is difficult to achieve with a sort based on extension.

I think an extension is a good idea and I can try writing it myself in a few weeks but I do need an overview of what is involved.

P.S. dirvish-subtree and the drop like indicators for it are very nice.

**Reply by alexluigit** [reply]
> I doesn't seem like that should need dired-revert.

It is being called  [here](https://github.com/Fuco1/dired-hacks/blob/master/dired-filter.el#L601)  for **every** filter (because all of them need a fully re-readin to work properly) in the filter-stack whenever the stack is changed. In the case of `dired-filter-group-mode`, let's say you have 5 entries in `dired-filter-group-saved-groups`, `dired-revert` is called 5 times on directory opening, which is rather expensive.

>  I used it to keep tex files at the top and then pdf files and then style files etc in a dired buffer

I agree that this feature is nice to have, but the price is too high under the architecture of  `dired-filter`. I'll see if I can come up with a better solution.

> I can try writing it myself in a few weeks but I do need an overview of what is involved.

I'll let you know if I write it before you do. Feel free to send a PR either on improving it or write it yourself, I'd happy to help.

**Reply by aikrahguzar** [reply]
> It is being called [here](https://github.com/Fuco1/dired-hacks/blob/master/dired-filter.el#L601) for **every** filter (because all of them need a fully re-readin to work properly) in the filter-stack whenever the stack is changed. In the case of `dired-filter-group-mode`, let's say you have 5 entries in `dired-filter-group-saved-groups`, `dired-revert` is called 5 times on directory opening, which is rather expensive.

Thanks a lot, I understand now. This is because `dired-filter-groups` works by repeated use of filter mechanism to partition files in two bins, one bin contains the files that match the filter and others that don't. The ones that don't are discarded and then we move to the next filter and indeed multiple passes are the issue here. 

But I think this can be avoided. Fundamentally all that is happening is a rearranging of lines in the buffer and this can be done in one pass, we go through each line, compute the first group it belongs too and put it in the bin for that group. This can be done either in buffer where we keep track of the location of where the next line for a group should be inserted using markers or using a temporary variable holding the strings for each group and these are just reinserted at the end. I don't know which way is better. This does not require reverting and I think should fix the one annoying aspect of the current implementation for me i.e if  a file matches two or more groups, it ends up under the last group which is the least important one for me.

 Since, I am unfamiliar with inner working of Dired, it might be that this single pass strategy can't work. If you see such a reason please let me know.

**Reply by alexluigit** [reply]
I've implemented a prototype for it as _dirvish-emerge.el_. I'm still working on it, but it should work for most of the cases. You can try out the new `dirvish-emerge-mode`, any feedback would be appreciated.

Just a glimpse.
![2022-06-27_22-27](https://user-images.githubusercontent.com/16313743/175965966-59691bf7-769f-4416-a956-cc63c09efdcd.png)

**Reply by aikrahguzar** [reply]
Thanks a lot, this looks very nice! I have written a more performant implementation of the `dired-filter-groups` and it works for my use case. My conclusion from writing it was basically that it is hard to get really good performance with a `predicate` based design and instead the best things to do is to write a function that goes through each line and spits out the groups. On the other hand predicate based design is much more user friendly.

From a very quick look at the code, I think you have found a good balance between the two approaches. I will try it out and report back!
Thanks a lot again for doing this.

**Reply by alexluigit** [reply]
Fixed a bug and improved the performance a bit in the latest commit. As of now, it takes ~1s to grouping a directory with 50000+ entries on my machine, while `dired-filter-group-mode` takes more than 3x time to apply the same filter.

> the best things to do is to write a function that goes through each line and spits out the groups

Indeed. This is a must. 

> My conclusion from writing it was basically that it is hard to get really good performance with a predicate based design

A predicate is expensive if it asks for `file-attributes` or anything relates disk I/O _on the fly_.  In other words, if we inhibit the predicate to access the resources it should not touch, then predicate/extensions/regex filters makes very little differences.  That's why I made the limitation to the predicates using `dirvish-emerge-define-predicate` and hook the grouping process onto `dirvish-setup-hook`.

**Reply by aikrahguzar** [reply]
I have played with it a little and overall it works better and quicker than what I had implemented. I hadn't though of using recently changed files as a separate group but it is actually what I wanted in a lot of cases so that is really good idea!

Here are some of the problems I have encountered:
1) There is a bug: the transient menu can only handle less than 9 groups. If `dirvish-emerge-groups` has more than 9 groups, I get `transient-setup: Key sequence 1 0 starts with non-prefix key 1`.

2) There doesn't seem like a way to apply the group automatically e.g. using a hook. I think calling `dirvish-emerge-mode` should apply the groups in `dirvish-emerge-groups`. I have tried adding `dirvish-emerge--apply` to the `dired-after-readin-hook` and even evaluating `(dirvish-emerge--apply)` manually but it doesn't works and I can only apply the groups using the transient menu.

3) I think some more predicates are needed, especially those that check for directories, executables and symlinks. 

I haven't looked at the code much but I can help with the predicates at least but I don't understand transient so unfortunately not there, at least anytime soon.

**Reply by alexluigit** [reply]
> the transient menu can only handle less than 9 groups

Thanks for reporting, I'll fix it ASAP.

> There doesn't seem like a way to apply the group automatically

```emacs-lisp
(add-hook 'dirvish-setup-hook 'dirvish-emerge-mode)
```

>  I can help with the predicates

Great! I believe `dirvish-emerge-define-predicate` is all you need.

**Reply by aikrahguzar** [reply]
> ```emacs-lisp
> (add-hook 'dirvish-setup-hook 'dirvish-emerge-mode)
> ```

This doesn't work for me. I still have to `M-x dirvish-emerge-menu RET RET` to actually apply the groups. Looking at the source code for `dirvish-emerge-mode`  it runs `dirvish-emerge--apply` but I even manually evaluating `dirvish-emerge--apply` doesn't help. I have to go through the transient menu.

> Great! I believe `dirvish-emerge-define-predicate` is all you need.

Yes, I was able to implement predicate for directories and symlinks pretty trivially with the help of `dirvish-emerge-define-predicate`,
```elisp
  (dirvish-emerge-define-predicate directories "Return t for directories."
    (eq t (file-attribute-type attrs)))
  (dirvish-emerge-define-predicate symlinks "Return t for symlinks."
    (stringp (file-attribute-type attrs)))
```
Implementing one for executables through file-attributes seems a bit more involved since that needs to know about the current user and their group. 

By the way, the behavior of predicates is confusing, it seems like they are matched in the order they appear in `dirvish-emerge--available-preds` while I would expect them to be match in the order in which they appear in `dirvish-emerge-groups`.

**Reply by alexluigit** [reply]
> I still have to M-x dirvish-emerge-menu RET RET to actually apply the groups.

That's weird. I can not repro this with `emacs -q`
```sh
emacs -q -L dirvish -l dirvish -L dirvish/extensions -l dirvish/extensions/dirvish-emerge.el --eval 
  "(progn (add-hook 'dirvish-setup-hook 'dirvish-emerge-mode) (dirvish \"~/test/\"))"
```

I have .dir-locals.el inside ~/test/ in this case. But it also works fine for me if I set `dirvish-emerge-groups` globally and open a random directory.

> while I would expect them to be match in the order in which they appear in dirvish-emerge-groups

I'm not able to repro this either. The order of the groups showing in the Dirvish buffer is always depended by `dirvish-emerge-groups`.

**Reply by aikrahguzar** [reply]
> That's weird. I can not repro this with `emacs -q`
> 
> ```shell
> emacs -q -L dirvish -l dirvish -L dirvish/extensions -l dirvish/extensions/dirvish-emerge.el --eval 
>   "(progn (add-hook 'dirvish-setup-hook 'dirvish-emerge-mode) (dirvish \"~/test/\"))"
> ```

This has disappeared for me too after updating! I think one of the recent commits fixed it so thanks a lot, you have doing amazing work.

> I'm not able to repro this either. The order of the groups showing in the Dirvish buffer is always depended by `dirvish-emerge-groups`.

This was a problem with the predicates I had written. The directory one doesn't check if the symlink resolves to a directory and I got confused, so sorry about that. I changed the directory predicate too,
```elisp
(dirvish-emerge-define-predicate directories "Return t for directories."
    (let ((type (file-attribute-type attrs)))
      (or (eq t type) (and (stringp type) (file-directory-p type)))))
```
and I have reproduced my old setup with some added niceties.

**Reply by aikrahguzar** [reply]
Another couple of suggestions:

1) It would be nice to be able to fold groups. Right now there is an option to hide a group but that also hides the headline for the group. I think it would be nicer to have it just hide the files with some way of indicating that the entries for this group are hidden. Similarly it would be nice to navigate by groups i.e to have jump to next/prev groups. This seems simple enough to be doable by me so if you would like it, I can do a pr this weekend.

2) Not related to `dirvish-emerge` but `dirvish` buffers always appear modified. This means for example that when I try to kill one I get prompted about it. I think normally `dired` buffers appear modified only if files have been marked/copied/moved etc. That seems like the better behavior to me. I was able to achieve the same for `dirvish` by an `around` advice to `dirvish-update-body-h` and `dirvish-emerge-mode` which wraps the calls to these functions with `with-silen-modifications`. This works fine for me but I am not sure it has no unintended consequences.

**Reply by aikrahguzar** [reply]
One last bit of magic from Dirvish, I had `dired-omit-mode` enabled through a hook in `dirvish-mode-hook`. Removing it from there and instead adding it to `dirvish-setup-hook` cut down the time it takes to finish loading a directory with about 15000 items (the biggest I know of on my system) from 2.1s to 0.4s which is almost instantaneous to me. Maybe it is worth mentioning somewhere in the documentation.

**Reply by alexluigit** [reply]
> I changed the directory predicate

Why not use the `type` argument directly?

```emacs-lisp
(dirvish-emerge-define-predicate directories 
  "Return t for directories."
  (eq (car type) 'dir))
```

Avoid I/O (like calling `file-directory-p`) in these predicates should bring some benefits to the performance, more importantly, it is critical on TRAMP connections.  In order to minimize the "choking" time, I guess `(equal (substring (file-attribute-modes attrs) 3 4) "x")` would be a better approach than `(file-executable-p full-name)` for now. Maybe we can come up with better solution than this in the future.

>  I think it would be nicer to have it just hide the files with some way of indicating that the entries for this group are hidden.

I agree. Maybe we can make the title always visible but displayed in different forms. 
Such as `- Recent ---` for unfolded and `- Recent [...] ---` for folded or something like that.

> This seems simple enough to be doable by me so if you would like it, I can do a pr this weekend.

Please feel free to do so.

> Not related to dirvish-emerge but dirvish buffers always appear modified. 

I've set `dired-clean-confirm-killing-deleted-buffers` to nil so this issue doesn't bother me at all, but I'll try out `with-silent-modifications`. Thanks for the suggestion.

> adding it to dirvish-setup-hook cut down the time it takes to finish loading a directory with about 15000 items ... Maybe it is worth mentioning somewhere in the documentation

Thanks for bringing this up. I'll update the doc

**Reply by aikrahguzar** [reply]
> Why not use the `type` argument directly?
> 
> ```emacs-lisp
> (dirvish-emerge-define-predicate directories 
>   "Return t for directories."
>   (eq (car type) 'dir))
> ```
This is indeed better, thanks! I think `type` argument wasn't available when I wrote it or maybe I just missed it. 
 
> Avoid I/O (like calling `file-directory-p`) in these predicates should bring some benefits to the performance, more importantly, it is critical on TRAMP connections. In order to minimize the "choking" time, I guess `(equal (substring (file-attribute-modes attrs) 3 4) "x")` would be a better approach than `(file-executable-p full-name)` for now. Maybe we can come up with better solution than this in the future.

This does seem like a good heuristic but it is different from `file-executable-p`. I looked to see if there was a way to figure out the current user and group from elisp but doesn't seem so. Maybe looking up envvars? I see a couple for user but not one for a group and I don't know if they are reliably set.

> Please feel free to do so.

Great! This is my weekend project then.

> Thanks for bringing this up. I'll update the doc

The situation with `dired-omit` is actually more complicated than I initially realized in two ways. 

1) The fast loading is due to the lazy loading magic `dirvish-does` so emacs is non-responsive for a bit (maybe a second or two in the case of directory with 15000 files) after the buffer is first rendered while `dired-omit-mode` and `dired-emerge-mode` do their stuff. This seems like the correct approach and showing something is much better than staring at an empty buffer and the result still appears pretty quickly even with a lot of files.

2) `dirvish-setup-hook` is actually not a good place to turn on minor modes that you might want to disable and I do want to be able to disable `dired-omit-mode` on occasion. But for the `dired-omit-mode` at least, this can be circumvented by
```emacs-lisp
(setq-default dired-omit-mode t)
(add-hook 'dirvish-setup-hook #'dired-omit-expunge)
```
This ensures that `dired-omit-mode` is on by default but can be disabled which is what I want.

**Reply by aikrahguzar** [reply]
Unrelated but there is also some weirdness with `dired-jump`. I have `dirvish-mode-line-position` set to `'disabled`. If I do `dired-jump` from a buffer `B` the dirvish buffer appears without any modeline as expected. If I now switch back to original buffer `B` and then do `dired-jump` again, the `dirvish-buffer` will appear with a modeline.

**Reply by alexluigit** [reply]
> I see a couple for user but not one for a group

Is it what you are looking for?
```emacs-lisp
(group-name (file-attribute-group-id attrs))
```

> The situation with dired-omit is actually more complicated than I initially realized ...

Personally I don't use this feature very often, does the following setup achieves the same results as `dired-omit-mode` does for you? 

```emacs-lisp
(setq dirvish-emerge-groups `(("Omit" (regex . ,dired-omit-files) t)))
```
If this works fine, maybe we can write a helper command to toggle the visibility of this "Omit" group. Basically the same thing as `dired-omit-mode`, but faster. 

> dirvish-mode-line-position

For now this option still only impacts full-screen sessions, so there might be some unexpected behaviors, I'll make it take effect in both scenarios.

**Reply by aikrahguzar** [reply]
> Is it what you are looking for?
> 
> ```emacs-lisp
> (group-name (file-attribute-group-id attrs))
> ```

This gives us the group the files belongs too. If the file doesn't have the `other` executable bit set, we need to know if the groups the current user belongs too to know if they can execute the file. This is probably going a bit too much into academic territory though. I haven't seen an executable without that bit set and `chmod +x` which is the most common way of making a file executable, sets all three bits.  So what you suggested along with the check that the file is not a directory should be enough i.e
```elisp
(dirvish-emerge-define-predicate executables "Matches executables."
    (and  (eq 'file (car type)) (eq ?x (aref (file-attribute-modes attrs) 3))))
```

> > The situation with dired-omit is actually more complicated than I initially realized ...
> 
> Personally I don't use this feature very often, does the following setup achieves the same results as `dired-omit-mode` does for you?
> 
> ```emacs-lisp
> (setq dirvish-emerge-groups `(("Omit" (regex . ,dired-omit-files) t)))
> ```
> 
> If this works fine, maybe we can write a helper command to toggle the visibility of this "Omit" group. Basically the same thing as `dired-omit-mode`, but faster.

I actually wanted to suggest that. I think toggle the visibility depending on the value of `dired-omit-mode` should work better than directly using `dired-omit-expunge`. But `dired-omit-files` is not the complete story, the correct way to get the omit regexp is `(dired-omit-regexp)`. This is because in addition to  `dired-omit-files` there is also `dired-omit-extensions`. The regexp `(dired-omit-regexp)` produces seemed to perform poorly compared to the one by dirvish-emerge for extensions so I checked the difference and it seems like the difference is dirvish checks for dot only once while the elements in `dired-omit-extensions` aren't really extensions and they are basically matched at the end of filename. I have only extensions in that variable (quite a lot of them) and my entry for an "Omitted" group is,
```elisp
("Omitted" (regex
            . ,(rx-to-string
                `(or (regexp ,dired-omit-files)
                     (and "."
                          (or . ,(mapcar (lambda (str) (substring str 1)) dired-omit-extensions))
                          eos)))))
```
The regexp above performs better but will fail for any entry in `dired-omit-extensions` that doesn't start with a dot.

> For now this option still only impacts full-screen sessions, so there might be some unexpected behaviors, I'll make it take effect in both scenarios.
Thanks for this.

**Reply by alexluigit** [reply]
> the elements in dired-omit-extensions aren't really extensions and they are basically matched at the end of filename

IMO that's not a good practice. Filenames like "foo~" should not be seen as file "foo" with "~" as its extension. So the "Omitted" group you provided is totally fine I guess. Maybe we can add a user option like `dirvish-emerge-omitted-files` and set its default value as you provided above.

**Reply by alexluigit** [reply]
I'm not sure if it's a bug of Dired, but after open a dired buffer in `emacs -q`, `(buffer-modified-p)` gives `t` instead of `nil`. 

>  I think normally dired buffers appear modified only if files have been marked/copied/moved etc.

This only happens after `dired-omit-mode` being applied to the buffer.

I hope ac08fc34ed7aa4d5aeaea9830cb1915d504f2820 handled the modification state correctly. Apart from using `with-silent-modifications`, I also gave every dirvish buffer an initial unmodified state.

**Reply by aikrahguzar** [reply]
> This only happens after `dired-omit-mode` being applied to the buffer.

That might be, I have had `dired-omit-mode` in my `dired-mode-hook` for a long time.

> I hope [ac08fc3](https://github.com/alexluigit/dirvish/commit/ac08fc34ed7aa4d5aeaea9830cb1915d504f2820) handled the modification state correctly. Apart from using `with-silent-modifications`, I also gave every dirvish buffer an initial unmodified state.

Yes, it works perfectly now. Thanks!

**Reply by aikrahguzar** [reply]
> IMO that's not a good practice. Filenames like "foo~" should not be seen as file "foo" with "~" as its extension. So the "Omitted" group you provided is totally fine I guess. Maybe we can add a user option like `dirvish-emerge-omitted-files` and set its default value as you provided above.

I agree with it not being a good practice but it seems to be common, e.g. doom uses that misfeature. That predicate is pretty problematic to provide by default, since it needs to be defined after `dired-omit-mode` is loaded and even then it won't change it the variables from `dired-omit-mode` are changed. I think it is better to put it in the documentation about how `dired-omit-mode` can be used with `dirvish-emerge`.

About avoiding `dired-omit-exounge`, I think the best way is to allow users to set the third entry of the members of `dired-emerge-groups`. This is already possible I think and just needs to be documented. We can allow three values for it, `nil` means the group is shown, `t` means that group is hidden but we show its header indicating that is it hidden, and if it is `omitted` then nothing is shown.

I can implement this in the pr I opened.
